### PR TITLE
[MV-431] [MV-MIOpen] Benchmark & Port TripletMarginLoss

### PR DESCRIPTION
**Pull request by long10024070**
* Added TripletMarginLoss foward and backward operations and kernels.
* Added driver test and gtest for TripletMarginLoss foward and backward operations.
* New API is guarded by MIOPEN_BETA_API macro.
* Compared to ROCm pytorch:

<details close>
<summary>float16</summary>
<br>

|      op_name      |  dtype  |    size    |   contiguous  | reduction |  model | direction | ROCm pytorch | MIOpen HIP | Improvement |
|:-----------------:|:-------:|:----------:|:-------------:|:---------:|:------:|:---------:|:------------:|:----------:|:-----------:|
| TripletMarginLoss | float16 |  [65536 1] |   contiguous  |    sum    |  dlrm  |    fwd    |     65034    |    33852   |     1.92    |
| TripletMarginLoss | float16 |  [65536 1] |   contiguous  |    sum    |  dlrm  |    bwd    |     44971    |    38021   |     1.18    |
| TripletMarginLoss | float16 |  [65536 1] | noncontiguous |    sum    |  dlrm  |    fwd    |     62898    |    33858   |     1.86    |
| TripletMarginLoss | float16 |  [65536 1] | noncontiguous |    sum    |  dlrm  |    bwd    |     44650    |    38009   |     1.17    |
| TripletMarginLoss | float16 |  [65536 1] |   contiguous  |    none   |  dlrm  |    fwd    |     48480    |    24237   |     2.00    |
| TripletMarginLoss | float16 |  [65536 1] |   contiguous  |    none   |  dlrm  |    bwd    |     47677    |    37775   |     1.26    |
| TripletMarginLoss | float16 |  [65536 1] | noncontiguous |    none   |  dlrm  |    fwd    |     48715    |    24257   |     2.01    |
| TripletMarginLoss | float16 |  [65536 1] | noncontiguous |    none   |  dlrm  |    bwd    |     47794    |    37756   |     1.27    |
| TripletMarginLoss | float16 |  [837 80]  |   contiguous  |    sum    |  yolor |    fwd    |     53661    |    38085   |     1.41    |
| TripletMarginLoss | float16 |  [837 80]  |   contiguous  |    sum    |  yolor |    bwd    |     59386    |    45335   |     1.31    |
| TripletMarginLoss | float16 |  [837 80]  | noncontiguous |    sum    |  yolor |    fwd    |     70712    |    38413   |     1.84    |
| TripletMarginLoss | float16 |  [837 80]  | noncontiguous |    sum    |  yolor |    bwd    |     75072    |    46313   |     1.62    |
| TripletMarginLoss | float16 |  [837 80]  |   contiguous  |    none   |  yolor |    fwd    |     49939    |    30816   |     1.62    |
| TripletMarginLoss | float16 |  [837 80]  |   contiguous  |    none   |  yolor |    bwd    |     61778    |    45844   |     1.35    |
| TripletMarginLoss | float16 |  [837 80]  | noncontiguous |    none   |  yolor |    fwd    |     58250    |    31171   |     1.87    |
| TripletMarginLoss | float16 |  [837 80]  | noncontiguous |    none   |  yolor |    bwd    |     73457    |    47787   |     1.54    |
| TripletMarginLoss | float16 | [800 4077] | noncontiguous |    sum    | yolov5 |    fwd    |    434657    |   294023   |     1.48    |
</details>

<details close>
<summary>float32</summary>
<br>

|      op_name      |  dtype  |    size    |   contiguous  | reduction |  model | direction | ROCm pytorch | MIOpen HIP | Improvement |
|:-----------------:|:-------:|:----------:|:-------------:|:---------:|:------:|:---------:|:------------:|:----------:|:-----------:|
| TripletMarginLoss | float32 |  [65536 1] |   contiguous  |    sum    |  dlrm  |    fwd    |     61302    |    37046   |     1.65    |
| TripletMarginLoss | float32 |  [65536 1] |   contiguous  |    sum    |  dlrm  |    bwd    |     44412    |    41185   |     1.08    |
| TripletMarginLoss | float32 |  [65536 1] | noncontiguous |    sum    |  dlrm  |    fwd    |     59329    |    36716   |     1.62    |
| TripletMarginLoss | float32 |  [65536 1] | noncontiguous |    sum    |  dlrm  |    bwd    |     43724    |    40715   |     1.07    |
| TripletMarginLoss | float32 |  [65536 1] |   contiguous  |    none   |  dlrm  |    fwd    |     43890    |    25435   |     1.73    |
| TripletMarginLoss | float32 |  [65536 1] |   contiguous  |    none   |  dlrm  |    bwd    |     42613    |    40474   |     1.05    |
| TripletMarginLoss | float32 |  [65536 1] | noncontiguous |    none   |  dlrm  |    fwd    |     43762    |    25906   |     1.69    |
| TripletMarginLoss | float32 |  [65536 1] | noncontiguous |    none   |  dlrm  |    bwd    |     42293    |    40117   |     1.05    |
| TripletMarginLoss | float32 |  [837 80]  |   contiguous  |    sum    |  yolor |    fwd    |     50833    |    42107   |     1.21    |
| TripletMarginLoss | float32 |  [837 80]  |   contiguous  |    sum    |  yolor |    bwd    |     61047    |    48437   |     1.26    |
| TripletMarginLoss | float32 |  [837 80]  | noncontiguous |    sum    |  yolor |    fwd    |     65593    |    43276   |     1.52    |
| TripletMarginLoss | float32 |  [837 80]  | noncontiguous |    sum    |  yolor |    bwd    |     72956    |    49467   |     1.47    |
| TripletMarginLoss | float32 |  [837 80]  |   contiguous  |    none   |  yolor |    fwd    |     42565    |    33001   |     1.29    |
| TripletMarginLoss | float32 |  [837 80]  |   contiguous  |    none   |  yolor |    bwd    |     60714    |    47565   |     1.28    |
| TripletMarginLoss | float32 |  [837 80]  | noncontiguous |    none   |  yolor |    fwd    |     57894    |    33285   |     1.74    |
| TripletMarginLoss | float32 |  [837 80]  | noncontiguous |    none   |  yolor |    bwd    |     72890    |    48310   |     1.51    |
| TripletMarginLoss | float32 | [800 4077] | noncontiguous |    none   | yolov5 |    fwd    |    392713    |   319432   |     1.23    |
</details>

<details close>
<summary>bfloat16</summary>
<br>

|      op_name      |   dtype  |    size    |   contiguous  | reduction |  model | direction | ROCm pytorch | MIOpen HIP | Improvement |
|:-----------------:|:--------:|:----------:|:-------------:|:---------:|:------:|:---------:|:------------:|:----------:|:-----------:|
| TripletMarginLoss | bfloat16 |  [65536 1] |   contiguous  |    sum    |  dlrm  |    fwd    |     72253    |    34144   |     2.12    |
| TripletMarginLoss | bfloat16 |  [65536 1] |   contiguous  |    sum    |  dlrm  |    bwd    |     51438    |    38398   |     1.34    |
| TripletMarginLoss | bfloat16 |  [65536 1] | noncontiguous |    sum    |  dlrm  |    fwd    |     69197    |    34119   |     2.03    |
| TripletMarginLoss | bfloat16 |  [65536 1] | noncontiguous |    sum    |  dlrm  |    bwd    |     50558    |    38416   |     1.32    |
| TripletMarginLoss | bfloat16 |  [65536 1] |   contiguous  |    none   |  dlrm  |    fwd    |     55342    |    24494   |     2.26    |
| TripletMarginLoss | bfloat16 |  [65536 1] |   contiguous  |    none   |  dlrm  |    bwd    |     53934    |    38079   |     1.42    |
| TripletMarginLoss | bfloat16 |  [65536 1] | noncontiguous |    none   |  dlrm  |    fwd    |     54701    |    24466   |     2.24    |
| TripletMarginLoss | bfloat16 |  [65536 1] | noncontiguous |    none   |  dlrm  |    bwd    |     53966    |    38216   |     1.41    |
| TripletMarginLoss | bfloat16 |  [837 80]  |   contiguous  |    sum    |  yolor |    fwd    |     58366    |    38314   |     1.52    |
| TripletMarginLoss | bfloat16 |  [837 80]  |   contiguous  |    sum    |  yolor |    bwd    |     68077    |    45703   |     1.49    |
| TripletMarginLoss | bfloat16 |  [837 80]  | noncontiguous |    sum    |  yolor |    fwd    |     69134    |    38680   |     1.79    |
| TripletMarginLoss | bfloat16 |  [837 80]  | noncontiguous |    sum    |  yolor |    bwd    |     81965    |    46648   |     1.76    |
| TripletMarginLoss | bfloat16 |  [837 80]  |   contiguous  |    none   |  yolor |    fwd    |     51070    |    31024   |     1.65    |
| TripletMarginLoss | bfloat16 |  [837 80]  |   contiguous  |    none   |  yolor |    bwd    |     70349    |    47070   |     1.49    |
| TripletMarginLoss | bfloat16 |  [837 80]  | noncontiguous |    none   |  yolor |    fwd    |     62910    |    31579   |     1.99    |
| TripletMarginLoss | bfloat16 |  [837 80]  | noncontiguous |    none   |  yolor |    bwd    |     84189    |    47273   |     1.78    |
| TripletMarginLoss | bfloat16 | [800 4077] | noncontiguous |    none   | yolov5 |    fwd    |    428386    |   290112   |     1.48    |
</details>

* Average over all cases:

|type|average|
|---|---|
|float16| 1.74 |
|float32| 1.56 |
|bfloat16| 1.86 |

### CODE REVIEW

File: driver/mloTripletMarginLossHost.hpp [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+           const T* I2,
+           const int p,
+           const float eps,
+           const tensor_view_t<2> I1_tv,
+           const tensor_view_t<2> I2_tv,
+           const size_t n,
+           const size_t C)
+{
+    // eps defined in
+    // https://pytorch.org/docs/stable/generated/torch.nn.PairwiseDistance.html
```
Comment (kyeonghwanryu) [inline driver/mloTripletMarginLossHost.hpp]: This comment is about the default value of eps. It's not here anymore so this is not the right spot.
And as far as I remember AMD didn't want external web link other than rocm environment to be in their code so I'll just remove it if I were you.

File: src/kernels/MIOpenLossSum.cpp
Comment (littlecutebird) [inline src/kernels/MIOpenLossSum.cpp:1]: Looks like MIOpenLossSum.cpp is duplicate of MIOpenLossReduce.cpp? But you removed REDUCE_SIZE in this file so REDUCE_SIZE is undefined.

File: driver/tripletmarginloss_driver.hpp [hunk 2]
```diff
@@ -0,0 +1,608 @@
...
+    dP       = std::vector<Tgpu>(positive_sz, static_cast<Tgpu>(0));
+    dN       = std::vector<Tgpu>(negative_sz, static_cast<Tgpu>(0));
+
+    outhost = std::vector<Tref>(out_sz, static_cast<Tref>(0));
+    dAhost  = std::vector<Tref>(anchor_sz, static_cast<Tref>(0));
+    dPhost  = std::vector<Tref>(positive_sz, static_cast<Tref>(0));
+    dNhost  = std::vector<Tref>(negative_sz, static_cast<Tref>(0));
+
+    for(int i = 0; i < anchor_sz; i++)
+        anchor[i] = prng::gen_A_to_B<Tgpu>(static_cast<Tgpu>(0.0), static_cast<Tgpu>(0.2));
```
Comment (kyeonghwanryu) [inline driver/tripletmarginloss_driver.hpp:301]: Is there any specific reason for these default ranges of values?

**Comment by kyeonghwanryu** [comment]
Sorry but I have to leave early for a company event. I'll resume the review tomorrow.

File: src/kernels/MIOpenLossSum.cpp
Comment (kyeonghwanryu) [reply src/kernels/MIOpenLossSum.cpp:1]: Seems like committing MIOpenLossSum.cpp was unintended because it's not in CMakeLists.txt. Please remove the unused one.

Comment (long10024070) [reply src/kernels/MIOpenLossSum.cpp:1]: I have removed MIOpenLossSum.cpp. And your are right, this file doesn't do anything as it is not listed in src/CMakeList.txt

File: driver/mloTripletMarginLossHost.hpp (hunk 1 above)
Comment (long10024070) [reply driver/mloTripletMarginLossHost.hpp]: I have removed those comments about `eps`.

File: src/include/miopen/tripletmarginloss/problem_description.hpp [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "Triplet Margin Loss: Output gradient tensor must have 1 dimension.");
+        if(!checkSameLength(aDesc, pDesc) || !checkSameLength(aDesc, nDesc))
+            MIOPEN_THROW(
+                miopenStatusBadParm,
+                "Triplet Margin Loss: Anchor, Positive, Negative tensor sizes do not match.");
+        return true;
+    }
+
+    bool IsReduced() const
```
Comment (kyeonghwanryu) [inline src/include/miopen/tripletmarginloss/problem_description.hpp]: Do you need both IsReduced and IsUnreduced? Aren't they exclusive?

File: src/tripletmarginloss/problem_description.cpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        return false;
+    for(int32_t i = 0; i < x.GetSize(); ++i)
+    {
+        if(x.GetStrides()[i] != y.GetStrides()[i])
+            return false;
+    }
+    return true;
+}
+
+NetworkConfig ForwardProblemDescription::MakeNetworkConfig() const
```
Comment (kyeonghwanryu) [inline src/tripletmarginloss/problem_description.cpp:67]: Please add strides info here.

File: driver/tripletmarginloss_driver.hpp (hunk 2 above)
Comment (long10024070) [reply driver/tripletmarginloss_driver.hpp:301]: Logically, no, it isn't.

But during my work with loss operations (reduction mode), I found that there are chance that the output value will be overflowed or meet floating point precision problem when sumed up or divided by a quite large number (above 10^5), if I use fp16 datatype, as largest normal number of fp16 is `65504` and smallest positive subnormal number is `0.000000059604645`.

That is why I have to make default ranges of values to be small, and adjust the ranges in different tensors, so when they subtract each other, which happends in most loss operations, the results will not too big to cause overflow when summed up. It also keeps the loss values of elements (before reduced) for not being too different, which will decrease the chance occuring floating point precision.

File: src/include/miopen/tripletmarginloss/problem_description.hpp (hunk 3 above)
Comment (long10024070) [reply src/include/miopen/tripletmarginloss/problem_description.hpp]: I believe that I need both of them, because there aren't any parts in my code validate the element size or dim[0] of output and grad output tensors. For example, an output tensor and an anchor tensor with shape (2) and (10,10) fail both `IsReduced` and `IsUnreduced` but they may pass all other checks.

File: src/tripletmarginloss/problem_description.cpp (hunk 4 above)
Comment (long10024070) [reply src/tripletmarginloss/problem_description.cpp:67]: Can you explain why strides info is needed? I only know that this information is used to determine whether to handle the operation in Contiguous or Noncontiguous ways. TripletMarginLoss considers both cases are the same. Both Contiguous and Noncontiguous cases are caculated with the same kernels.

Comment (kyeonghwanryu) [reply src/tripletmarginloss/problem_description.cpp:67]: Yeah my mistake. In most cases there are separate kernels and it's a common mistake, so I left a mechanical comment. In this case you don't need to do that.

File: src/include/miopen/tripletmarginloss/problem_description.hpp (hunk 3 above)
Comment (kyeonghwanryu) [reply src/include/miopen/tripletmarginloss/problem_description.hpp]: Okay. You're using them in IsApplicable. But if there's !IsReduced() && !IsUnreduced() case, doesn't it need to be considered as an invalid input?
That kinds of checks should be done at the constructor of `ProblemDescription`.
Check this review comment. https://github.com/ROCm/MIOpen/pull/2623#discussion_r1450730047

And if you do so, I think you can use one of IsReduced and IsUnreduced and remove other one.

**Comment by kyeonghwanryu** [comment]
Looks good to me other than current comments.

File: src/include/miopen/tripletmarginloss/problem_description.hpp (hunk 3 above)
Comment (long10024070) [reply src/include/miopen/tripletmarginloss/problem_description.hpp]: I have moved those checks which contain `MIOPEN_THROW` to problem description (forward/backward) constructor. It includes those in `IsSameType()` and `IsRightLength()` which made me removed those 2 functions along with `IsUnreduced()`.

File: driver/tripletmarginloss_driver.hpp (hunk 2 above)
Comment (kyeonghwanryu) [reply driver/tripletmarginloss_driver.hpp:301]: Ok I understand that.

**Comment by kyeonghwanryu** [comment]
Good work!
